### DISCUSSION TITLE
How is 'upright' property  defined?

### DISCUSSION THREAD

**Discussion by JankowskiKamil** [jsvine/pdfplumber#642] [Q&A]
Hello,

I use this amazing library and sometimes have a problem with vertical / not upright chars. Is there any chance to get chars rotation value? I suspect that in cases where upright property is miscalculated, the rotation is equal to 89.99999... not 90. 

I try to find the problem and wonder if this value is based on: upright = True if rotation == 90 else False, but I can't find the line of code responsible for this.

Unfortunately, I can't send a PDF file

**Comment by jsvine** [comment]
Hi @JankowskiKamil, and great question. The `upright` property comes directly from `pdfminer.six`, the dependency we use to get char/rect/etc.-level information. Specifically, it is computed here: https://github.com/pdfminer/pdfminer.six/blob/43c8fc8557528463c99598049b7005ae96ab8084/pdfminer/layout.py#L379

As you'll see from that part of the code, `upright` is computed from object's `self.matrix` property and a local variable `scaling`:

```python
        (a, b, c, d, e, f) = self.matrix
        self.upright = 0 < a * d * scaling and b * c <= 0
```

The `self.matrix` property is more straightforward, since `pdfminer.six` exposes it and `pdfplumber` could extract it. In fact, PR https://github.com/jsvine/pdfplumber/pull/100 proposed this a while ago, and I mentioned adding this as a possibility in https://github.com/jsvine/pdfplumber/issues/368#issuecomment-798808061. Due to time constraints, I haven't yet done this. But your question reminds me that it would be a useful thing, and I'll see if I can do something about it.

**Reply by jsvine** [reply]
Update: Just pushed an attempt at making the current transformation matrix (CTM) accessible for characters, and introduced a class (`pdfplumber.ctm.CTM`) for calculating its scale, skew, and translation: https://github.com/jsvine/pdfplumber/commit/ae6f99e691de72987ab8166403a39776a41d4c30

It turns out that CTMs achieve rotation by combining scale and skew. In most standard cases, it seems, rotation will be equal to the x-axis skew. So, to calculate the rotation, you can run something like this:

```python
from pdfplumber.ctm import CTM
my_char = pdf.pages[0].chars[3]
my_char_ctm = CTM(*my_char["matrix"])
my_char_rotation = my_char_ctm.skew_x
```

To test this out, you can install the unreleased code via `pip install -e git+https://github.com/jsvine/pdfplumber.git@feature-ctm`

If you get the chance, please let me know whether it suits your particular situation.

**Reply by JankowskiKamil** [reply]
oh sorry, I didn't see this answer and i forgot the code a long time ago, but I think the first one helped me a lot :D
